File: workers/python/kafka_worker.py
import asyncio
import json
import os
import sys
from datetime import datetime, timezone
from aiokafka import AIOKafkaConsumer
from sqlalchemy import create_engine, text

# add 프로젝트 루트를 sys.path에 추가하여 base_queue 임포트 가능하게 함
sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..'))
# from workers.python.base_queue import OrderEvent

# DB 연결 정보
DB_URL = os.getenv("DATABASE_URL", "postgresql://user:password@localhost:5432/mq_db")
engine = create_engine(DB_URL)

async def consume_orders():
    bootstrap_servers = os.getenv("KAFKA_BOOTSTRAP_SERVERS", "localhost:9092")
    group_id = "payment-group"
    
    consumer = AIOKafkaConsumer(
        "orders",
        bootstrap_servers=bootstrap_servers,
        group_id=group_id,
        value_deserializer=lambda v: json.loads(v.decode('utf-8')),
        auto_offset_reset="earliest"
    )
    
    await consumer.start()
    print(f"[*] Python Worker ({group_id}) started. Listening on {bootstrap_servers}...")
    
    try:
        async for msg in consumer:
            event_dict = msg.value
            order_id = event_dict.get("order_id")
            published_at_str = event_dict.get("published_at")
            
            # 지연 시간 계산 (ms)
            consumed_at = datetime.now(timezone.utc)
            published_at = datetime.fromisoformat(published_at_str.replace("Z", "+00:00"))
            latency_ms = int((consumed_at - published_at).total_seconds() * 1000)
            
            print(f" [x] {group_id} Received: {order_id} (Latency: {latency_ms}ms)")
            
            # DB 저장
            with engine.connect() as conn:
                query = text("""
                    INSERT INTO processed_events (event_id, group_id, mq_type, data, latency_ms)
                    VALUES (:event_id, :group_id, :mq_type, :data, :latency_ms)
                """)
                conn.execute(query, {
                    "event_id": order_id,
                    "group_id": group_id,
                    "mq_type": "kafka",
                    "data": json.dumps(event_dict),
                    "latency_ms": latency_ms
                })
                conn.commit()
                
    finally:
        await consumer.stop()

if __name__ == "__main__":
    asyncio.run(consume_orders())

File: workers/node/src/kafka.worker.ts
import { Kafka } from 'kafkajs';
import { Client } from 'pg';
import type { ProcessedEvent } from '@mq/shared-node';

const DB_URL = process.env['DATABASE_URL'] ?? 'postgresql://postgres:password@localhost:5432/mq_db';
const KAFKA_BOOTSTRAP_SERVERS = process.env['KAFKA_BOOTSTRAP_SERVERS'] ?? 'localhost:9092';
const GROUP_ID = 'inventory-group';

async function insertProcessedEvent(
    client: Client,
    event: Omit<ProcessedEvent, 'id' | 'processedAt'>
): Promise<void> {
    const query = `
        INSERT INTO processed_events (event_id, group_id, mq_type, data, latency_ms)
        VALUES ($1, $2, $3, $4, $5)
    `;
    await client.query(query, [
        event.eventId,
        event.groupId,
        event.mqType,
        JSON.stringify(event.data),
        event.latencyMs,
    ]);
}

async function runWorker() {
    // 1. DB 연결
    const dbClient = new Client({ connectionString: DB_URL });
    await dbClient.connect();
    console.log(`Connected to DB for ${GROUP_ID}`);

    // 2. Kafka 연결
    const kafka = new Kafka({
        clientId: 'inventory-worker',
        brokers: [KAFKA_BOOTSTRAP_SERVERS],
    });

    const consumer = kafka.consumer({ groupId: GROUP_ID });
    await consumer.connect();
    await consumer.subscribe({ topic: 'orders', fromBeginning: true });

    console.log(`[*] Node.js Worker (${GROUP_ID}) started. Listening on ${KAFKA_BOOTSTRAP_SERVERS}...`);

    await consumer.run({
        eachMessage: async ({ message }: { topic: string; partition: number; message: { value: Buffer | null } }) => {
            if (!message.value) return;

            const eventData = JSON.parse(message.value.toString()) as Record<string, unknown>;
            const orderId = (eventData['orderId'] ?? eventData['order_id']) as string;
            const publishedAtStr = (eventData['publishedAt'] ?? eventData['published_at']) as string;

            const consumedAt = new Date();
            const publishedAt = new Date(publishedAtStr);
            const latencyMs = consumedAt.getTime() - publishedAt.getTime();

            console.log(` [x] ${GROUP_ID} Received: ${orderId} (Latency: ${latencyMs}ms)`);

            await insertProcessedEvent(dbClient, {
                eventId: orderId,
                groupId: GROUP_ID,
                mqType: 'kafka',
                data: eventData,
                latencyMs,
            });
        },
    });
}

runWorker().catch(err => {
    console.error('Error running Node.js Worker:', err);
    process.exit(1);
});
